safe_transaction_service/history/models.py

```diff
@@ -645,55 +593,79 @@ def ether_incoming_txs_for_address(self, address: str):
+            "_token_id",
+            "token_address",
+        ]
+        erc20_queryset = ERC20Transfer.objects.token_txs()
+        erc721_queryset = ERC721Transfer.objects.token_txs()
         return (
-            EthereumEvent.objects.erc20_and_721_events()
-            .annotate(
-                to=RawSQL("arguments->>%s", ("to",)),  # Order is really important!
-                _from=RawSQL("arguments->>%s", ("from",)),
-                value=RawSQL("(arguments->>%s)::numeric", ("value",)),
-                transaction_hash=F("ethereum_tx_id"),
-                block_number=F("ethereum_tx__block_id"),
-                execution_date=F("ethereum_tx__block__timestamp"),
-                token_id=RawSQL("(arguments->>%s)::numeric", ("tokenId",)),
-                token_address=F("address"),
-            )
-            .order_by("-ethereum_tx__block_id")
-        )
-
-    def token_txs_for_address(self, address: str):
-        return self.token_txs().filter(
-            Q(arguments__to=address) | Q(arguments__from=address)
+            erc20_queryset.values(*values)
+            .union(erc721_queryset.values(*values))
+            .order_by("-block_number")
         )
 
     def token_incoming_txs_for_address(self, address: str):
-        return self.token_txs().filter(arguments__to=address)
+        values = [
+            "block_number",
+            "transaction_hash",
+            "to",
+            "_from",
+            "_value",
+            "execution_date",
+            "_token_id",
+            "token_address",
+        ]
+        erc20_queryset = ERC20Transfer.objects.incoming(address).token_txs()
+        erc721_queryset = ERC721Transfer.objects.incoming(address).token_txs()
+        return (
+            erc20_queryset.values(*values)
+            .union(erc721_queryset.values(*values))
+            .order_by("-block_number")
+        )
 
     def ether_and_token_txs(self, address: str):
-        tokens_queryset = self.token_txs_for_address(address)
+        erc20_queryset = ERC20Transfer.objects.to_or_from(address).token_txs()
+        erc721_queryset = ERC721Transfer.objects.to_or_from(address).token_txs()
         ether_queryset = self.ether_txs_for_address(address)
-        return self.union_ether_and_token_txs(tokens_queryset, ether_queryset)
+        return self.union_ether_and_token_txs(
+            erc20_queryset, erc721_queryset, ether_queryset
+        )
 
     def ether_and_token_incoming_txs(self, address: str):
-        tokens_queryset = self.token_incoming_txs_for_address(address)
+        erc20_queryset = ERC20Transfer.objects.incoming(address).token_txs()
+        erc721_queryset = ERC721Transfer.objects.incoming(address).token_txs()
         ether_queryset = self.ether_incoming_txs_for_address(address)
-        return self.union_ether_and_token_txs(tokens_queryset, ether_queryset)
+        return self.union_ether_and_token_txs(
+            erc20_queryset, erc721_queryset, ether_queryset
+        )
```

Comment:
Isn't it possible to use the `QuerySet` returned by `token_incoming_txs_for_address`? The only set missing would be the ether transfers.

Reply:
Where?

Reply:
Ok, I get you, you are right but I prefer to leave this o a previous optimization, as we will be sorting twice